### PR TITLE
retrybp: Implement retry-after logic in CappedExponentialBackoff

### PR DESCRIPTION
**Pull request by fishy**
Define retrybp.RetryAfterError interface and add support to it to
CappedExponentialBackoff, and implement it in httpbp as
httpbp.ClientError.

### CODE REVIEW

**Comment by fishy** [comment]
Note: before this change we mainly focused on http server in `httpbp` package, and the added APIs to `httpbp` package are focused on the client side. I'm not sure if we want to support them as our API. If we don't feel like support them, I can move that into an example to retrybp.RetryAfterError instead.

File: httpbp/errors.go [hunk 1]
```diff
@@ -633,3 +638,112 @@ func ErrorForCode(code int) *ErrorResponse {
...
+// The caller can (optionally) choose to read the body and feed that back into
+// ClientError.AdditionalInfo (it's always empty string when returned), e.g.:
+//
+//     resp, err := http.Get("https://www.reddit.com")
+//     if err != nil {
+//       return err
+//     }
+//     defer httpbp.DrainAndClose(resp.Body)
+//     err = httpbp.ClientErrorFromResponse(resp)
+//     if err != nil && errors.As(err, &ce) {
```
Comment (pacejackson) [inline httpbp/errors.go:695]: `ce` doesn't exist yet in this example.

Comment (fishy) [reply httpbp/errors.go:695]: :doh: I forgot to delete the `&& ...` part when splitting the if.

**Review by konradreiche — APPROVED** [review]
Generally fine with supporting this as part of our API since in reality many of our code bases will need this and I am not sure how else to incorporate this code.

**Review by pacejackson — APPROVED** [review]

File: httpbp/errors.go (hunk 1 above)
Comment (fishy) [reply httpbp/errors.go]: Actually I moved this to testable example so it's guaranteed to be correct code :)

**Pull request merged**